### DISCUSSION TITLE
DacFx provisioning support

### DISCUSSION THREAD

**Discussion by bjorkstromm** [microsoft/aspire#1696] [Ideas]
I played with DacFx support yesterday, and wondered if a pull request would be of interest? If not, are there plans for a `aspire-contrib` repo/organization, where a curated set of extensions and and components could live?

Context:
https://x.com/bjorkstromm/status/1747749746272288877?s=20

High-level functionality:
- Possibility to deploy `dacpac` using Aspire, example code
```csharp
using Aspire.Hosting.DacFx;

var builder = DistributedApplication.CreateBuilder(args);

builder.AddDacFxProvisioning();

builder.AddSqlServerContainer("sqlserver")
    .AddDatabase("database")
    .WithDacPackage<Projects.Database>();

builder.Build().Run();
```
- `AddDacFxProvisioning()` registers a lifecycle hook with a dac provisioner, that will start dacpac deployments after endpoints are allocated.
- The dac provisioner scans the resources and searches for the SQL Databases (both Azure SQL and SQL Server container) which are annotated with dacpac information (annotation is done using `WithDacPackage<T>()`).
- The dac provisioner kicks off background work and starts deploying dac-packages to the databases.

Utilizing [Microsoft.Build.Sql](https://github.com/microsoft/DacFx) we can get the dacpac directly via project reference in the AppHost. In order to get the output path, I needed to modify the `CreateServiceMetadataSources` target to also include OutputPath. Adding support for pointing at the dacpac-file directly should also be considered.

**Comment by dzsquared** [comment]
From the DacFx side, we're interested in helping folks deploy the database changes within their Aspire host's configuration!

Should you need anything from DacFx (or Microsoft.Build.Sql), please don't hesitate to flag us down over on https://github.com/microsoft/DacFx in case I miss a notification here.

**Reply by bjorkstromm** [reply]
Thanks! First thing I think we'd need to solve is this https://github.com/dotnet/aspire/issues/2511

Without this we won't get the path of the `dacpac` for a referenced `sqlproj`.

**Comment by hjrb** [comment]
It would be greate to have this feature. I am a DB first believer

**Comment by jmezach** [comment]
We are working on something similar for `MSBuild.Sdk.SqlProj`, see https://github.com/rr-wfm/MSBuild.Sdk.SqlProj/pull/585 for more details.
